File: teacher.js
// Конфигурация
const API_URL = "https://api.newlxp.ru/graphql";
const EMAIL = "evloevam@magas.ithub.ru";
const PASSWORD = "1Q2w3a4e$#";

// Функция для выполнения запросов GraphQL
async function graphqlRequest(query, variables, token = null) {
  const headers = {
    "Content-Type": "application/json",
  };
  if (token) {
    headers.Authorization = `Bearer ${token}`;
  }

  const response = await fetch(API_URL, {
    method: "POST",
    headers,
    body: JSON.stringify({
      query,
      variables,
    }),
  });

  if (!response.ok) {
    throw new Error(`HTTP error! status: ${response.status}`);
  }

  const result = await response.json();

  if (result.errors) {
    console.error("GraphQL errors:", result.errors);
    throw new Error(`GraphQL Error: ${result.errors[0]?.message || "Unknown error"}`);
  }

  return result.data;
}

// Авторизация
async function signIn() {
  const query = `
    query SignIn($input: SignInInput!) {
      signIn(input: $input) {
        user {
          id
          isLead
          __typename
        }
        accessToken
        __typename
      }
    }
  `;

  const variables = {
    input: {
      email: EMAIL,
      password: PASSWORD,
    },
  };

  const data = await graphqlRequest(query, variables);
  const token = data.signIn.accessToken;
  const userId = data.signIn.user.id;
  console.log("Авторизация успешна.");
  return { token, userId };
}

// Получение дисциплин, которые вы преподаёте
async function getMyDisciplines(token) {
  const query = `
    query GetTeacherDisciplines {
      getMe {
        teacher {
          assignedDisciplines_V2 {
            isActivated
            disciplineId
            discipline {
              id
              name
              code
              archivedAt
              suborganization {
                id
                name
                organization {
                  id
                  name
                }
              }
              studyPeriods {
                id
                name
                status
                startDate
                endDate
              }
            }
          }
        }
      }
    }
  `;

  const data = await graphqlRequest(query, {}, token);
  return data.getMe.teacher.assignedDisciplines_V2;
}

// Получение структуры дисциплины (главы и темы) - ИСПРАВЛЕННЫЙ ЗАПРОС ПО ОРИГИНАЛУ
async function getDisciplineStructure(token, disciplineId, userId) {
  const query = `
    query GetDisciplineDataWithChaptersById($input: GetDisciplineByIdInput!, $userId: UUID!, $userRole: Identity_RoleType) {
      getDisciplineById(input: $input) {
        id
        maxScore
        name
        studyHoursCount
        archivedAt
        isAutoMeetingLink
        academicDifferenceDiscipline {
          id
          completionStatus
          __typename
        }
        suborganizationId
        retake {
          completionStatus
          contentFillingTeacherId
          content(userId: $userId, userRole: $userRole) {
            id
            __typename
          }
          __typename
        }
        chapters {
          description
          id
          name
          templateDisciplineChapterId
          topics {
            id
            isCheckPoint
            isUneditable
            isForPortfolio
            maxScore
            methodologicalType
            name
            order
            studyHoursCount
            chapterId
            templateDisciplineTopicId
            content {
              blocks {
                ... on InfoDisciplineTopicContentBlock {
                  id
                  kind
                  __typename
                }
                ... on TestDisciplineTopicContentBlock {
                  id
                  kind
                  __typename
                }
                ... on TaskDisciplineTopicContentBlock {
                  id
                  kind
                  __typename
                }
                __typename
              }
            }
            __typename
          }
          discipline {
            maxScore
            studyHoursCount
            templateDiscipline {
              name
              __typename
            }
            __typename
          }
          __typename
        }
        __typename
      }
    }
  `;

  const variables = {
    input: { disciplineId: disciplineId },
    userId: userId,
    userRole: "TEACHER", // Указываем вашу роль
  };

  const data = await graphqlRequest(query, variables, token);
  return data.getDisciplineById;
}

// Основная функция
async function main() {
  try {
    const { token, userId } = await signIn();

    const disciplinesData = await getMyDisciplines(token);

    if (!disciplinesData || disciplinesData.length === 0) {
      console.log("У вас нет назначенных дисциплин или произошла ошибка.");
      return;
    }

    console.log("\n--- Ваши дисциплины ---");
    const activeDisciplines = disciplinesData.filter(item => item.isActivated && !item.discipline.archivedAt);
    activeDisciplines.forEach((item, index) => {
      const disc = item.discipline;
      console.log(`${index + 1}. ${disc.name} (код: ${disc.code})`);
      console.log(`   ID: ${disc.id}`);
      console.log(`   Организация: ${disc.suborganization.organization.name}`);
      console.log(`   Подорганизация: ${disc.suborganization.name}`);
      // Проверяем учебные периоды
      if (disc.studyPeriods && disc.studyPeriods.length > 0) {
        const currentPeriod = disc.studyPeriods.find(p => p.status === 'STARTED' || p.status === 'PLANNED');
        if (currentPeriod) {
          console.log(`   Учебный период: ${currentPeriod.name} (${currentPeriod.status})`);
        }
      }
      console.log("-".repeat(20));
    });

    if (activeDisciplines.length === 0) {
      console.log("У вас нет активных (не архивных) дисциплин.");
      return;
    }

    // Для чтения ввода в Node.js нужно использовать readline или стороннюю библиотеку
    // Используем readline для примера
    const readline = require('readline');
    const rl = readline.createInterface({
      input: process.stdin,
      output: process.stdout
    });

    const question = (query) => new Promise((resolve) => rl.question(query, resolve));

    let choice;
    while (true) {
      const answer = await question(`\nВыберите номер дисциплины (1-${activeDisciplines.length}): `);
      choice = parseInt(answer) - 1;
      if (choice >= 0 && choice < activeDisciplines.length) {
        break;
      } else {
        console.log("Неверный номер. Попробуйте снова.");
      }
    }
    rl.close(); // Закрываем интерфейс readline после получения ввода

    const selectedDiscipline = activeDisciplines[choice];
    const selectedDisciplineId = selectedDiscipline.discipline.id;
    const selectedDisciplineName = selectedDiscipline.discipline.name;

    console.log(`\n--- Загрузка структуры дисциплины: ${selectedDisciplineName} ---`);
    const disciplineStructure = await getDisciplineStructure(token, selectedDisciplineId, userId);

    if (!disciplineStructure) {
      console.log("Не удалось получить структуру дисциплины.");
      return;
    }

    console.log(`\nДисциплина: ${disciplineStructure.name}`);
    console.log(`Всего часов: ${disciplineStructure.studyHoursCount}`);
    console.log(`Макс. баллов: ${disciplineStructure.maxScore}`);
    console.log(`Архивирована: ${disciplineStructure.archivedAt ? 'Да' : 'Нет'}\n`);

    disciplineStructure.chapters.forEach(chapter => {
      console.log(`--- Глава: ${chapter.name} ---`);
      console.log(`    Описание: ${chapter.description || 'Нет'}`);
      let totalChapterHours = 0;
      chapter.topics
        .sort((a, b) => a.order - b.order) // Сортировка тем по порядку
        .forEach(topic => {
          const checkpointStr = topic.isCheckPoint ? " (Контрольная точка)" : "";
          // Исправленная строка с форматированием номера
          console.log(`  ${(topic.order + 1).toString().padStart(2, '0')}. ${topic.name}${checkpointStr} - ${topic.studyHoursCount} ч. (Тип: ${topic.methodologicalType})`);
          totalChapterHours += topic.studyHoursCount;
        });
      console.log(`  Итого по главе: ${totalChapterHours} ч.\n`);
    });

  } catch (error) {
    console.error("Произошла ошибка:", error.message);
  }
}

// Запуск основной функции
main();

File: teacher.py
import requests
import json

API_URL = "https://api.newlxp.ru/graphql"
EMAIL = "evloevam@magas.ithub.ru"
PASSWORD = "1Q2w3a4e$#"

# Авторизация
def sign_in():
    query = """
    query SignIn($input: SignInInput!) {
      signIn(input: $input) {
        user {
          id
          isLead
          __typename
        }
        accessToken
        __typename
      }
    }
    """
    variables = {
        "input": {
            "email": EMAIL,
            "password": PASSWORD
        }
    }
    response = requests.post(API_URL, json={"query": query, "variables": variables})
    if response.status_code == 200:
        data = response.json()
        if 'errors' in data:
            print(f"Ошибка GraphQL при авторизации: {data['errors']}")
            exit(1)
        token = data["data"]["signIn"]["accessToken"]
        user_id = data["data"]["signIn"]["user"]["id"]
        print(f"Авторизация успешна.")
        return token, user_id
    else:
        print(f"Ошибка HTTP при авторизации: {response.status_code}")
        print(response.text)
        exit(1)

# Получение дисциплин, которые вы преподаёте
def get_my_disciplines(token):
    query = """
    query GetTeacherDisciplines {
        getMe {
            teacher {
                assignedDisciplines_V2 {
                    isActivated
                    disciplineId
                    discipline {
                        id
                        name
                        code
                        archivedAt
                        suborganization {
                            id
                            name
                            organization {
                                id
                                name
                            }
                        }
                        studyPeriods {
                            id
                            name
                            status
                            startDate
                            endDate
                        }
                    }
                }
            }
        }
    }
    """

    headers = {
        "Authorization": f"Bearer {token}",
        "Content-Type": "application/json"
    }

    response = requests.post(API_URL, json={"query": query}, headers=headers)
    if response.status_code == 200:
        data = response.json()
        if 'errors' in data:
            print(f"Ошибка GraphQL при получении дисциплин: {data['errors']}")
            return []
        return data["data"]["getMe"]["teacher"]["assignedDisciplines_V2"]
    else:
        print(f"Ошибка HTTP при получении дисциплин: {response.status_code}")
        print(response.text)
        return []

# Получение структуры дисциплины (главы и темы) - ИСПРАВЛЕННЫЙ ЗАПРОС ПО ОРИГИНАЛУ
def get_discipline_structure(token, discipline_id, user_id):
    query = """
    query GetDisciplineDataWithChaptersById($input: GetDisciplineByIdInput!, $userId: UUID!, $userRole: Identity_RoleType) {
      getDisciplineById(input: $input) {
        id
        maxScore
        name
        studyHoursCount
        archivedAt
        isAutoMeetingLink
        academicDifferenceDiscipline {
          id
          completionStatus
          __typename
        }
        suborganizationId
        retake {
          completionStatus
          contentFillingTeacherId
          content(userId: $userId, userRole: $userRole) {
            id
            __typename
          }
          __typename
        }
        chapters {
          description
          id
          name
          templateDisciplineChapterId
          topics {
            id
            isCheckPoint
            isUneditable
            isForPortfolio
            maxScore
            methodologicalType
            name
            order
            studyHoursCount
            chapterId
            templateDisciplineTopicId
            content {
              blocks {
                ... on InfoDisciplineTopicContentBlock {
                  id
                  kind
                  __typename
                }
                ... on TestDisciplineTopicContentBlock {
                  id
                  kind
                  __typename
                }
                ... on TaskDisciplineTopicContentBlock {
                  id
                  kind
                  __typename
                }
                __typename
              }
            }
            __typename
          }
          discipline {
            maxScore
            studyHoursCount
            templateDiscipline {
              name
              __typename
            }
            __typename
          }
          __typename
        }
        __typename
      }
    }
    """
    # Передаем disciplineId внутри объекта input
    # userId и userRole передаем как отдельные переменные
    variables = {
        "input": {"disciplineId": discipline_id},
        "userId": user_id,
        "userRole": "TEACHER" # Указываем вашу роль
    }

    headers = {
        "Authorization": f"Bearer {token}",
        "Content-Type": "application/json"
    }

    response = requests.post(API_URL, json={"query": query, "variables": variables}, headers=headers)
    if response.status_code == 200:
        data = response.json()
        # Проверяем на наличие ошибок GraphQL в ответе
        if 'errors' in data:
            print(f"Ошибка GraphQL при получении структуры дисциплины: {data['errors']}")
            return None
        return data["data"]["getDisciplineById"]
    else:
        print(f"Ошибка HTTP при получении структуры дисциплины: {response.status_code}")
        print(response.text)
        return None

def main():
    token, user_id = sign_in()

    disciplines_data = get_my_disciplines(token)

    if not disciplines_data:
        print("У вас нет назначенных дисциплин или произошла ошибка.")
        return

    print("\n--- Ваши дисциплины ---")
    active_disciplines = [item for item in disciplines_data if item['isActivated'] and not item['discipline']['archivedAt']]
    for idx, item in enumerate(active_disciplines):
        disc = item['discipline']
        print(f"{idx + 1}. {disc['name']} (код: {disc['code']})")
        print(f"   ID: {disc['id']}")
        print(f"   Организация: {disc['suborganization']['organization']['name']}")
        print(f"   Подорганизация: {disc['suborganization']['name']}")
        # Проверяем учебные периоды
        if disc['studyPeriods']:
            current_period = next((p for p in disc['studyPeriods'] if p['status'] in ['STARTED', 'PLANNED']), None)
            if current_period:
                 print(f"   Учебный период: {current_period['name']} ({current_period['status']})")
        print("-" * 20)

    if not active_disciplines:
        print("У вас нет активных (не архивных) дисциплин.")
        return

    while True:
        try:
            choice = int(input(f"\nВыберите номер дисциплины (1-{len(active_disciplines)}): ")) - 1
            if 0 <= choice < len(active_disciplines):
                selected_discipline = active_disciplines[choice]
                break
            else:
                print("Неверный номер. Попробуйте снова.")
        except ValueError:
            print("Пожалуйста, введите число.")

    selected_discipline_id = selected_discipline['discipline']['id']
    selected_discipline_name = selected_discipline['discipline']['name']

    print(f"\n--- Загрузка структуры дисциплины: {selected_discipline_name} ---")
    discipline_structure = get_discipline_structure(token, selected_discipline_id, user_id)

    if not discipline_structure:
        print("Не удалось получить структуру дисциплины.")
        return

    print(f"\nДисциплина: {discipline_structure['name']}")
    print(f"Всего часов: {discipline_structure['studyHoursCount']}")
    print(f"Макс. баллов: {discipline_structure['maxScore']}")
    print(f"Архивирована: {'Да' if discipline_structure['archivedAt'] else 'Нет'}\n")

    for chapter in discipline_structure['chapters']:
        print(f"--- Глава: {chapter['name']} ---")
        print(f"    Описание: {chapter['description'] if chapter['description'] else 'Нет'}")
        total_chapter_hours = 0
        for topic in sorted(chapter['topics'], key=lambda x: x['order']):
            checkpoint_str = " (Контрольная точка)" if topic['isCheckPoint'] else ""
            print(f"  {topic['order'] + 1:2d}. {topic['name']}{checkpoint_str} - {topic['studyHoursCount']} ч. (Тип: {topic['methodologicalType']})")
            total_chapter_hours += topic['studyHoursCount']
            # Выводим типы блоков контента (если нужно)
            # if topic['content']['blocks']:
            #     block_types = [block['kind'] for block in topic['content']['blocks']]
            #     print(f"       Блоки: {', '.join(block_types)}")
        print(f"  Итого по главе: {total_chapter_hours} ч.\n")


if __name__ == "__main__":
    main()
